File: Part2/Section 5 - Multiple Linear Regression/Regresión_lineal_multiple.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Mon Nov 18 09:29:49 2019

@author: root
"""
#regresión lineal múltiple

import  numpy as np
import matplotlib.pyplot as plt
import pandas as pd

#importar datase
df=pd.read_csv("50_Startups.csv") 

x=df.iloc[:,:-1].values
y=df.iloc[:, 4].values

#datos  categóricos no ordinales
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import make_column_transformer 
onehotencoder=make_column_transformer((OneHotEncoder(),[3]),remainder="passthrough")
x=onehotencoder.fit_transform(x)

#evitar la trampa de las variables ficticias
x=x[:,1:]

#dividir dataset en conjunto de entrenamiento y testing
from sklearn.model_selection import train_test_split as splitter
x_train, x_test, y_train,  y_test = splitter(x, y, test_size=0.2, random_state=0)

#Modelo de Regresión de entrenamiento
from sklearn.linear_model import LinearRegression
regression=LinearRegression()
regression.fit(x_train, y_train)

#prediccion de resultado en test
y_pred=regression.predict(x_test)

#eliminar variables hacia atras para optimizar el modelo (manual)
"""import statsmodels.regression.linear_model as stat
x=np.append(arr =np.ones((50,1)).astype(int), values =x, axis=1)
sl=0.05
x_opt=x[:,[0,1,2,3,4,5]].tolist()
regression_ols=stat.OLS(endog=y, exog=x_opt).fit()
print(regression_ols.summary())

x_opt=x[:,[0,2,3,4,5]].tolist()
regression_ols=stat.OLS(endog=y, exog=x_opt).fit()
print(regression_ols.summary())
  
x_opt=x[:,[0,3,4,5]].tolist()
regression_ols=stat.OLS(endog=y, exog=x_opt).fit()
print(regression_ols.summary())
"""
#eliminar variables para optimizar el modelo con P valor y R Cuadrado
"""
import statsmodels.regression.linear_model as sm
def backwardElimination(x, SL):    
    numVars = len(x[0])    
    temp = np.zeros((50,6)).astype(int)    
    for i in range(0, numVars):        
        regressor_OLS = sm.OLS(endog=y, exog=x_opt).fit()      
        maxVar = max(regressor_OLS.pvalues).astype(float)        
        adjR_before = regressor_OLS.rsquared_adj.astype(float)        
        if maxVar > SL:            
            for j in range(0, numVars - i):                
                if (regressor_OLS.pvalues[j].astype(float) == maxVar):                    
                    temp[:,j] = x[:, j]                    
                    x = np.delete(x, j, 1)                    
                    tmp_regressor = sm.OLS(y, x.tolist()).fit()                    
                    adjR_after = tmp_regressor.rsquared_adj.astype(float)                    
                    if (adjR_before >= adjR_after):                        
                        x_rollback = np.hstack((x, temp[:,[0,j]]))                        
                        x_rollback = np.delete(x_rollback, j, 1)     
                        print (regressor_OLS.summary())                        
                        return x_rollback                    
                    else:                        
                        continue    
    print(regressor_OLS.summary())    
    return x 
x=np.append(arr =np.ones((50,1)).astype(int), values =x, axis=1) 
SL = 0.05
x_opt = x[:, [0, 1, 2, 3, 4, 5]].tolist()
x_Modeled = backwardElimination(x_opt, SL)
"""    

File: Part2/Section 5 - Multiple Linear Regression/Regresión_lineal_multiple.R
#Regresión Lineal Múltiple
#Plantilla para el Pre Procesado de Datos
#Importar Dataset
df=read.csv('50_Startups.csv')
#df=dataset[,2:3]

#codificar variables categóricas

df$State=factor(df$State,
                   levels=c("New York", "California", "Florida"),
                   labels=c(1,2,3))

#Dividir los datos en entrenamiento y conjunto de test
#install.packages("caTools")
library(caTools)
set.seed(123)
sample=sample.split(df$Profit, SplitRatio = 0.8)
training_set=subset(df, sample==TRUE)
testing_set=subset(df, sample==FALSE)

#Ajustar el modelo de Regresión Lineal Múltiple con el conjunto de entrenamiento
regression=lm(formula = Profit ~.,
              data=training_set)
#predecir resultados con testing dataset
y_pred=predict(regression,newdata=testing_set)

#construir modelo optimo  cpn la Eliminación Hacia Atrás (manual)
sl = 0.05
regression=lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend + State, 
              data=df)
summary(regression)

regression=lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend , 
              data=df)
summary(regression)

regression=lm(formula = Profit ~ R.D.Spend + Marketing.Spend , 
              data=df)
summary(regression)
regression=lm(formula = Profit ~ R.D.Spend, 
              data=df)
summary(regression)  

# #construir modelo optimo  cpn la Eliminación Hacia Atrás (auto)
# backwardElimination <- function(x, sl) {
#   numVars = length(x)
#   for (i in c(1:numVars)){
#     regressor = lm(formula = Profit ~ ., data = x)
#     maxVar = max(coef(summary(regressor))[c(2:numVars), "Pr(>|t|)"])
#     if (maxVar > sl){
#       j = which(coef(summary(regressor))[c(2:numVars), "Pr(>|t|)"] == maxVar)
#       x = x[, -j]
#     }
#     numVars = numVars - 1
#   }
#   return(summary(regressor))
# }
# 
# sl = 0.05
# dataset = df[, c(1,2,3,4)]
# backwardElimination(training_set, sl)


